### PR TITLE
Added caching of pofiles

### LINKED ISSUE #62: Add a cache option
Cache would be stored in like `~/.potodo/cache.json`, and it would contain the last refresh date and refresh every month or something ? The reservations would still be updated everytime the command is ran

### PR DESCRIPTION
**Pull request by Seluj78**
Closes #62

### CODE REVIEW

File: potodo/_cache.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import pickle
+from datetime import datetime
+from datetime import timedelta
+from typing import Optional
+from typing import Tuple
+
+
+def _get_cache_file_content(
+    path: str = ".potodo/cache.pickle",
+) -> Optional[Tuple[datetime, dict]]:
```
Comment (JulienPalard) [inline potodo/_cache.py:11]: À première vue c'est plutôt un `Tuple[Optional[datetime], Optional[dict]]` ou un `Union[Tuple[None, None], Tuple[datetime, dict]]`

File: potodo/potodo.py [hunk 2]
```diff
@@ -121,14 +125,33 @@ def exec_potodo(
...
+
+    if (
+        no_cache
+        or check_output(["git", "status", "--porcelain"], encoding="utf-8") != ""
+    ):
+        update_cache = True
+    else:
+        dt_expiry, content = _get_cache_file_content()
+        if content:
+            if dt_expiry < datetime.utcnow():
```
Comment (JulienPalard) [inline potodo/potodo.py:144]: Est-ce que ça ne serait pas plus lisible de bouger ce if dans `_get_cache_file_content` ? Elle renverai ainsi soit `None` soit `data`. Et ce serait plus symétrique car c'est `_set_cache_file_content` qui stocke cette date.

**Comment by JulienPalard** [comment]
If you want to avoid displaying stale cache (which would happen if you `git pull` between two invocations of potodo), you could store the mtime of the file when it's cached, and compare the mtime of the file when it's read: if the mtime is different, the file has changed it should be read again.

**Comment by Seluj78** [comment]
> If you want to avoid displaying stale cache (which would happen if you `git pull` between two invocations of potodo), you could store the mtime of the file when it's cached, and compare the mtime of the file when it's read: if the mtime is different, the file has changed it should be read again.

The problem is, I store the raw `po_files_and_dirs` dictionnary, which i then directly used in the rest of the process. See https://github.com/Seluj78/Potodo/pull/68/files#diff-0dc0200b5c248313b31e1a7f9570054dR169

That would mean adding another layer of running through the files and directories one by one and adding their `mtime` manually to the dict somewhere and then removing it when the cache is validated

**Comment by JulienPalard** [comment]
> The problem is, I store the raw `po_files_and_dirs` dictionnary, which i then directly used in the rest of the process. See https://github.com/Seluj78/Potodo/pull/68/files#diff-0dc0200b5c248313b31e1a7f9570054dR169

I didn't said it was straightforward :p

I think it still could be done, in `get_po_files_from_repo`, when you create each `PoFileStats(po_file)`, you could have a
 "if is in cache and mtime not changed, get from cache, else compute and cache" thing.

So out of `get_po_files_from_repo` you can have a mixed set from both the cache and fresh data, instead of it being all white or all black.

**Comment by Seluj78** [comment]
NICE
`potodo -p /Users/seluj78/Projects/python-docs-fr --no-cache  4.97s user 0.56s system 75% cpu 7.321 total`
`potodo -p /Users/seluj78/Projects/python-docs-fr  2.26s user 0.49s system 68% cpu 4.008 total`

File: potodo/potodo.py [hunk 3]
```diff
@@ -376,6 +376,13 @@ def main() -> None:
         help="show issue creation dates",
     )
 
+    parser.add_argument(
+        "--no-cache",
+        action="store_true",
+        dest="no_cache",
+        help="Disables cache (Cache is disabled when files are modified in git)",
```
Comment (JulienPalard) [inline potodo/potodo.py:383]:
```suggestion
        help="Disables cache (Cache is disabled when files are modified)",
```

File: tox.ini [hunk 4]
```diff
@@ -15,7 +15,7 @@ commands = flake8 tests/ potodo/
 
 [testenv:black]
 deps = black
-commands = black --check --diff tests/ potodo/
+commands = black --check --diff tests/ potodo/ --line-length 88
```
Comment (JulienPalard) [inline tox.ini]:
```suggestion
commands = black --check --diff tests/ potodo/
```
Not needed, 88 is the default.

File: potodo/potodo.py [hunk 5]
```diff
@@ -121,14 +118,13 @@ def exec_potodo(
     :param exclude_reserved: Will print out only files that aren't reserved
     :param only_reserved: Will print only reserved fils
     :param show_reservation_dates: Will show the reservation dates
+    :param no_cache: Disables cache (Cache is disabled when files are modified in git)
```
Comment (JulienPalard) [inline potodo/potodo.py:121]:
```suggestion
    :param no_cache: Disables cache (Cache is disabled when files are modified)
```

**Review by JulienPalard — CHANGES_REQUESTED** [review]
LGTM with nits.

**Review by JulienPalard — APPROVED** [review]

**Pull request merged**